src/ssl.c

```diff
@@ -33089,6 +33236,70 @@ int wolfSSL_EC_POINT_set_affine_coordinates_GFp(const WOLFSSL_EC_GROUP *group,
 #if !defined(WOLFSSL_ATECC508A) && !defined(WOLFSSL_ATECC608A) && \
     !defined(HAVE_SELFTEST)
 #if !defined(HAVE_FIPS) || (defined(HAVE_FIPS_VERSION) && (HAVE_FIPS_VERSION>2))
+int wolfSSL_EC_POINT_add(const WOLFSSL_EC_GROUP *group, WOLFSSL_EC_POINT *r,
+                         const WOLFSSL_EC_POINT *p1,
+                         const WOLFSSL_EC_POINT *p2, WOLFSSL_BN_CTX *ctx)
+{
+    mp_int a, prime;
+    mp_digit mp = 0;
+    int ret = WOLFSSL_FAILURE;
+
+    (void)ctx;
+
+    if (!group || !r || !p1 || !p2) {
+        WOLFSSL_MSG("wolfSSL_EC_POINT_add error");
+        return WOLFSSL_FAILURE;
+    }
+
+    if (setupPoint(r) != WOLFSSL_SUCCESS ||
+        setupPoint(p1) != WOLFSSL_SUCCESS ||
+        setupPoint(p2) != WOLFSSL_SUCCESS) {
+        WOLFSSL_MSG("setupPoint error");
+        return WOLFSSL_FAILURE;
+    }
+
+    /* read the curve prime and a */
+    if (mp_init_multi(&prime, &a, NULL, NULL, NULL, NULL) != MP_OKAY) {
+        WOLFSSL_MSG("mp_init_multi error");
+        goto cleanup;
+    }
+
+    if (mp_read_radix(&a, ecc_sets[group->curve_idx].Af, MP_RADIX_HEX)
+            != MP_OKAY) {
+        WOLFSSL_MSG("mp_read_radix a error");
+        goto cleanup;
+    }
+
+    if (mp_read_radix(&prime, ecc_sets[group->curve_idx].prime, MP_RADIX_HEX)
+            != MP_OKAY) {
+        WOLFSSL_MSG("mp_read_radix prime error");
+        goto cleanup;
+    }
+
+    if (mp_montgomery_setup(&prime, &mp) != MP_OKAY) {
+        WOLFSSL_MSG("mp_montgomery_setup nqm error");
+        goto cleanup;
+    }
+
+    if (ecc_projective_add_point((ecc_point*)p1->internal, (ecc_point*)p2->internal,
```

Comment:
The points being added must be in Montgomery form.
Is this always true?

Reply:
I'm not sure. I based the `wolfSSL_EC_POINT_add` implementation on other uses of `ecc_projective_add_point`. Is there some way I could check if they are in Montgomery form?

Reply:
The point won't have its ordinates in Montgomery form.
You'll need to convert to and from in this function.

To convert to Montgomery, take a look at: ecc.c:2635-2684 (wc_ecc_mulmod_ex())
Map the point back call: ecc_map_ex()

Add a test and maybe even try compiling a simplified test against OpenSSL and wolfSSL and see if you get the same results.

Reply:
Does this mean that something similar should be done in `wolfSSL_EC_POINT_mul`?

Reply:
Yes, well, sort of.
ecc_mul2add does everything internally.
The wc_ecc_mulmod shouldn't be mapping the results back to affine and out of Montgomery form. Last parameter is 1 and should be 0 for both calls. The add is expecting Montgomery form. Then the mapping done after (ecc_map) converts the Montgomery form in projective to affine in normal form.
The others cases look right.